File: heat3d_mpi.c
#include "common.h"
#include <mpi.h>

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);

    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // Распределение по Z
    int base_nz = NZ / size;
    int start_z = rank * base_nz;
    int end_z = (rank == size - 1) ? NZ : start_z + base_nz;
    int local_nz = end_z - start_z;
    int local_nz_with_halo = local_nz + 2;

    if (rank == 0) {
        printf("MPI C version: %d processes\n", size);
        printf("Global grid: %dx%dx%d, Steps: %d\n", NX, NY, NZ, NUM_STEPS);
    }

    // Выделение памяти в порядке [k][j][i]
    double* temp = malloc(sizeof(double) * local_nz_with_halo * NY * NX);
    double* new_temp = malloc(sizeof(double) * local_nz_with_halo * NY * NX);

    if (temp == NULL || new_temp == NULL) {
        printf("Process %d: Memory allocation failed\n", rank);
        MPI_Abort(MPI_COMM_WORLD, 1);
    }

    // Инициализируем все точки boundary temperature (20.0)
    for (int k = 0; k < local_nz_with_halo; k++) {
        for (int j = 0; j < NY; j++) {
            for (int i = 0; i < NX; i++) {
                int idx = k * NY * NX + j * NX + i;
                temp[idx] = BOUNDARY_TEMP;
                new_temp[idx] = BOUNDARY_TEMP;
            }
        }
    }

    // Инициализация горячего куба (без halo)
    int center_x = NX/2, center_y = NY/2, center_z = NZ/2;
    for (int k_local = 1; k_local <= local_nz; k_local++) {
        int k_global = start_z + (k_local - 1);
        for (int j = 0; j < NY; j++) {
            for (int i = 0; i < NX; i++) {
                if (abs(i - center_x) < CUBE_SIZE && 
                    abs(j - center_y) < CUBE_SIZE && 
                    abs(k_global - center_z) < CUBE_SIZE) {
                    int idx = k_local * NY * NX + j * NX + i;
                    temp[idx] = INITIAL_TEMP;
                    new_temp[idx] = INITIAL_TEMP;
                }
            }
        }
    }

    // Коэффициенты
    double dx2 = DX * DX;
    double dy2 = DY * DY;
    double dz2 = DZ * DZ;
    double factor = DT * ALPHA;

    double start_time = MPI_Wtime();

    for (int step = 0; step < NUM_STEPS; step++) {
        // Обмен halo-областями
        if (size > 1) {
            MPI_Request requests[4];
            int request_count = 0;
            
            // Отправляем нижний слой предыдущему процессу, получаем верхний halo
            if (rank > 0) {
                MPI_Isend(&temp[1 * NY * NX], NY * NX, MPI_DOUBLE, rank - 1, 0, MPI_COMM_WORLD, &requests[request_count++]);
                MPI_Irecv(&temp[0 * NY * NX], NY * NX, MPI_DOUBLE, rank - 1, 1, MPI_COMM_WORLD, &requests[request_count++]);
            }
            
            // Отправляем верхний слой следующему процессу, получаем нижний halo
            if (rank < size - 1) {
                MPI_Isend(&temp[local_nz * NY * NX], NY * NX, MPI_DOUBLE, rank + 1, 1, MPI_COMM_WORLD, &requests[request_count++]);
                MPI_Irecv(&temp[(local_nz + 1) * NY * NX], NY * NX, MPI_DOUBLE, rank + 1, 0, MPI_COMM_WORLD, &requests[request_count++]);
            }
            
            // Ждем завершения всех операций
            if (request_count > 0) {
                MPI_Waitall(request_count, requests, MPI_STATUSES_IGNORE);
            }
        }

        // Вычисления - только внутренние точки
        for (int k_local = 1; k_local <= local_nz; k_local++) {
            for (int j = 1; j < NY-1; j++) {
                for (int i = 1; i < NX-1; i++) {
                    int idx = k_local * NY * NX + j * NX + i;
                    int idx_xp = k_local * NY * NX + j * NX + (i+1);
                    int idx_xm = k_local * NY * NX + j * NX + (i-1);
                    int idx_yp = k_local * NY * NX + (j+1) * NX + i;
                    int idx_ym = k_local * NY * NX + (j-1) * NX + i;
                    int idx_zp = (k_local+1) * NY * NX + j * NX + i;
                    int idx_zm = (k_local-1) * NY * NX + j * NX + i;

                    double d2x = (temp[idx_xp] - 2*temp[idx] + temp[idx_xm]) / dx2;
                    double d2y = (temp[idx_yp] - 2*temp[idx] + temp[idx_ym]) / dy2;
                    double d2z = (temp[idx_zp] - 2*temp[idx] + temp[idx_zm]) / dz2;
                    
                    new_temp[idx] = temp[idx] + factor * (d2x + d2y + d2z);
                }
            }
        }

        // Копирование обновленных значений
        for (int k_local = 1; k_local <= local_nz; k_local++) {
            for (int j = 1; j < NY-1; j++) {
                for (int i = 1; i < NX-1; i++) {
                    int idx = k_local * NY * NX + j * NX + i;
                    temp[idx] = new_temp[idx];
                }
            }
        }

        // Граничные условия Дирихле на глобальных границах X и Y
        for (int k = 0; k < local_nz_with_halo; k++) {
            for (int j = 0; j < NY; j++) {
                for (int i = 0; i < NX; i++) {
                    int idx = k * NY * NX + j * NX + i;
                    // Границы по X
                    if (i == 0 || i == NX-1) {
                        temp[idx] = BOUNDARY_TEMP;
                        new_temp[idx] = BOUNDARY_TEMP;
                    }
                    // Границы по Y
                    if (j == 0 || j == NY-1) {
                        temp[idx] = BOUNDARY_TEMP;
                        new_temp[idx] = BOUNDARY_TEMP;
                    }
                }
            }
        }

        // Глобальные граничные условия по Z для крайних процессов
        if (rank == 0) {
            for (int j = 0; j < NY; j++) {
                for (int i = 0; i < NX; i++) {
                    int idx = 0 * NY * NX + j * NX + i;
                    temp[idx] = BOUNDARY_TEMP;
                    new_temp[idx] = BOUNDARY_TEMP;
                }
            }
        }
        if (rank == size - 1) {
            for (int j = 0; j < NY; j++) {
                for (int i = 0; i < NX; i++) {
                    int idx = (local_nz + 1) * NY * NX + j * NX + i;
                    temp[idx] = BOUNDARY_TEMP;
                    new_temp[idx] = BOUNDARY_TEMP;
                }
            }
        }

        // Вывод прогресса
        if (rank == 0 && step % OUTPUT_INTERVAL == 0) {
            double elapsed = MPI_Wtime() - start_time;
            printf("Step %d/%d, Time: %.2fs\n", step, NUM_STEPS, elapsed);
        }
    }

    double total_time = MPI_Wtime() - start_time;

    // Сохранение результатов
    if (rank == 0) {
        FILE* file = fopen("mpi_results.bin", "wb");
        if (!file) {
            printf("Error opening file for writing\n");
            MPI_Abort(MPI_COMM_WORLD, 1);
        }

        // Создаем глобальную сетку в порядке [i][j][k]
        double* global_grid = malloc(NX * NY * NZ * sizeof(double));
        
        // Копируем данные процесса 0
        for (int k_local = 0; k_local < local_nz; k_local++) {
            for (int j = 0; j < NY; j++) {
                for (int i = 0; i < NX; i++) {
                    int src_idx = (k_local + 1) * NY * NX + j * NX + i;
                    int dst_idx = i * NY * NZ + j * NZ + (start_z + k_local);
                    global_grid[dst_idx] = temp[src_idx];
                }
            }
        }

        // Получаем данные от других процессов
        for (int proc = 1; proc < size; proc++) {
            int proc_base_nz = NZ / size;
            int proc_start = proc * proc_base_nz;
            int proc_nz = (proc == size - 1) ? (NZ - proc_start) : proc_base_nz;
            
            double* proc_data = malloc(proc_nz * NY * NX * sizeof(double));
            MPI_Recv(proc_data, proc_nz * NY * NX, MPI_DOUBLE, proc, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
            
            for (int k_local = 0; k_local < proc_nz; k_local++) {
                for (int j = 0; j < NY; j++) {
                    for (int i = 0; i < NX; i++) {
                        int src_idx = k_local * NY * NX + j * NX + i;
                        int dst_idx = i * NY * NZ + j * NZ + (proc_start + k_local);
                        global_grid[dst_idx] = proc_data[src_idx];
                    }
                }
            }
            free(proc_data);
        }

        // Сохраняем 3D-сетку в бинарный файл
        for (int i = 0; i < NX; i++) {
            for (int j = 0; j < NY; j++) {
                fwrite(&global_grid[i * NY * NZ + j * NZ], sizeof(double), NZ, file);
            }
        }
        
        fclose(file);
        free(global_grid);
        printf("Results saved to mpi_results.bin\n");
        
    } else {
        // Отправляем данные процессу 0 (без halo)
        double* send_data = malloc(local_nz * NY * NX * sizeof(double));
        
        for (int k_local = 0; k_local < local_nz; k_local++) {
            for (int j = 0; j < NY; j++) {
                for (int i = 0; i < NX; i++) {
                    int src_idx = (k_local + 1) * NY * NX + j * NX + i;
                    int dst_idx = k_local * NY * NX + j * NX + i;
                    send_data[dst_idx] = temp[src_idx];
                }
            }
        }
        
        MPI_Send(send_data, local_nz * NY * NX, MPI_DOUBLE, 0, 0, MPI_COMM_WORLD);
        free(send_data);
    }

    if (rank == 0) {
        printf("MPI C (%d processes): %.2f sec\n", size, total_time);
    }

    free(temp);
    free(new_temp);
    MPI_Finalize();
    return 0;
}

File: heat3d_mpi.py
from mpi4py import MPI
import numpy as np
import time

# Константы
NX, NY, NZ = 450, 450, 450
NUM_STEPS = 1000
DT = 0.01
ALPHA = 0.00001172
DX, DY, DZ = 0.001, 0.001, 0.001
OUTPUT_INTERVAL = 100

CUBE_SIZE = 90
INITIAL_TEMP = 1000.0
BOUNDARY_TEMP = 20.0

def solve_heat_mpi():
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    # Распределение по Z
    base_nz = NZ // size
    start_z = rank * base_nz
    end_z = NZ if rank == size - 1 else start_z + base_nz
    local_nz = end_z - start_z
    local_nz_with_halo = local_nz + 2

    if rank == 0:
        print(f"MPI Python FIXED: {size} processes")
        print(f"Grid: {NX}x{NY}x{NZ}, Steps: {NUM_STEPS}")

    # Выделение памяти в порядке [k][j][i]
    local_temp = np.empty((local_nz_with_halo, NY, NX), dtype=np.float64)
    local_new  = np.empty_like(local_temp)

    # Инициализируем все точки boundary temperature
    local_temp[:] = BOUNDARY_TEMP
    local_new[:]  = BOUNDARY_TEMP

    # Инициализация горячего куба
    center_x = NX // 2
    center_y = NY // 2
    center_z = NZ // 2

    i = np.arange(NX)[None, None, :]
    j = np.arange(NY)[None, :, None]
    k_global = np.arange(start_z, end_z)[:, None, None]

    hot_mask = (
        (np.abs(i - center_x) < CUBE_SIZE) &
        (np.abs(j - center_y) < CUBE_SIZE) &
        (np.abs(k_global - center_z) < CUBE_SIZE)
    )

    # Помещаем горячий блок в local_temp[1:local_nz+1]
    local_temp[1:local_nz+1] = np.where(
        hot_mask, INITIAL_TEMP, local_temp[1:local_nz+1]
    )
    local_new[1:local_nz+1] = local_temp[1:local_nz+1]

    # Коэффициенты
    dx2 = DX * DX
    dy2 = DY * DY
    dz2 = DZ * DZ
    factor = DT * ALPHA

    start_time = time.time()

    # Основной цикл
    for step in range(NUM_STEPS):

        # Обмен halo-областями
        if size > 1:

            # Отправляем нижний слой предыдущему процессу, получаем верхний halo
            if rank > 0:
                sendbuf = np.ascontiguousarray(local_temp[1, :, :])
                recvbuf = np.full_like(sendbuf, BOUNDARY_TEMP)
                comm.Sendrecv(
                    sendbuf=sendbuf, dest=rank - 1, sendtag=0,
                    recvbuf=recvbuf, source=rank - 1, recvtag=1
                )
                local_temp[0, :, :] = recvbuf
            else:
                local_temp[0, :, :] = BOUNDARY_TEMP

            # Отправляем верхний слой следующему процессу, получаем нижний halo
            if rank < size - 1:
                sendbuf2 = np.ascontiguousarray(local_temp[local_nz, :, :])
                recvbuf2 = np.full_like(sendbuf2, BOUNDARY_TEMP)
                comm.Sendrecv(
                    sendbuf=sendbuf2, dest=rank + 1, sendtag=1,
                    recvbuf=recvbuf2, source=rank + 1, recvtag=0
                )
                local_temp[local_nz + 1, :, :] = recvbuf2
            else:
                local_temp[local_nz + 1, :, :] = BOUNDARY_TEMP

        else:
            local_temp[0, :, :] = BOUNDARY_TEMP
            local_temp[-1, :, :] = BOUNDARY_TEMP

        # Вычисления - только внутренние точки
        core = local_temp[1:-1, 1:-1, 1:-1]

        d2x = (local_temp[1:-1, 1:-1, 2:] -
               2 * core +
               local_temp[1:-1, 1:-1, :-2]) / dx2

        d2y = (local_temp[1:-1, 2:, 1:-1] -
               2 * core +
               local_temp[1:-1, :-2, 1:-1]) / dy2

        d2z = (local_temp[2:, 1:-1, 1:-1] -
               2 * core +
               local_temp[:-2, 1:-1, 1:-1]) / dz2

        local_new[1:-1, 1:-1, 1:-1] = core + factor * (d2x + d2y + d2z)
        local_temp[1:-1, 1:-1, 1:-1] = local_new[1:-1, 1:-1, 1:-1]

        # Граничные условия Дирихле на глобальных границах X и Y
        local_temp[:, :, 0]      = BOUNDARY_TEMP
        local_temp[:, :, NX - 1] = BOUNDARY_TEMP
        local_temp[:, 0, :]      = BOUNDARY_TEMP
        local_temp[:, NY - 1, :] = BOUNDARY_TEMP

        # Глобальные граничные условия по Z для крайних процессов
        if rank == 0:
            local_temp[0, :, :] = BOUNDARY_TEMP
        if rank == size - 1:
            local_temp[local_nz + 1, :, :] = BOUNDARY_TEMP

        # Вывод прогресса
        if rank == 0 and step % OUTPUT_INTERVAL == 0:
            elapsed = time.time() - start_time
            print(f"Step {step}/{NUM_STEPS}, Time: {elapsed:.2f}s")

    total_time = time.time() - start_time

    # Сохранение результатов
    if rank == 0:

        global_grid = np.empty((NX, NY, NZ), dtype=np.float64)

        # Копируем данные процесса 0
        global_grid[:, :, :local_nz] = local_temp[1:local_nz+1].transpose(2, 1, 0)

        # Получаем данные от других процессов
        for proc in range(1, size):
            proc_base_nz = NZ // size
            proc_start = proc * proc_base_nz
            proc_nz = NZ - proc_start if proc == size - 1 else proc_base_nz

            buf = np.empty((proc_nz, NY, NX), dtype=np.float64)
            comm.Recv(buf, source=proc, tag=0)

            global_grid[:, :, proc_start:proc_start+proc_nz] = buf.transpose(2, 1, 0)

        # Сохраняем 3D-сетку в бинарный файл
        with open("python_mpi_results.bin", "wb") as f:
            for i in range(NX):
                for j in range(NY):
                    global_grid[i, j, :].tofile(f)

        print(f"MPI Python ({size} processes): {total_time:.2f} sec")
        print("Saved to python_mpi_results.bin")

    else:
        # Отправляем данные процессу 0 (без halo)
        sendbuf = np.ascontiguousarray(local_temp[1:local_nz + 1])
        comm.Send(sendbuf, dest=0, tag=0)


if __name__ == "__main__":
    solve_heat_mpi()